src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsTag.cs

```diff
@@ -0,0 +1,167 @@
+
+        public InlineDiagnosticsTag(string errorType, DiagnosticData diagnostic, IEditorFormatMap editorFormatMap, InlineDiagnosticsLocations location)
+            : base(editorFormatMap)
+        {
+            ErrorType = errorType;
+            _diagnostic = diagnostic;
+            Location = location;
+        }
+
+        /// <summary>
+        /// Creates a GraphicsResult object which is the error block based on the geometry and formatting set for the item.
+        /// </summary>
+        public override GraphicsResult GetGraphics(IWpfTextView view, Geometry bounds, TextFormattingRunProperties format)
+        {
+            var block = new TextBlock
+            {
+                FontFamily = format.Typeface.FontFamily,
+                FontSize = 0.75 * format.FontRenderingEmSize,
+                FontStyle = FontStyles.Normal,
+                Foreground = format.ForegroundBrush,
+                Padding = new Thickness(left: 2, top: 0, right: 2, bottom: 0),
+            };
+
+            var id = new Run(_diagnostic.Id);
+            var link = new Hyperlink(id)
+            {
+                NavigateUri = new Uri(_diagnostic.HelpLink)
+            };
+
+            link.RequestNavigate += HandleRequestNavigate;
+            block.Inlines.Add(link);
+            block.Inlines.Add(": " + _diagnostic.Message);
+            block.Measure(new Size(double.PositiveInfinity, double.PositiveInfinity));
+            block.Arrange(new Rect(block.DesiredSize));
+
+            var lineHeight = Math.Floor(format.Typeface.FontFamily.LineSpacing * block.FontSize);
+            var image = new CrispImage
+            {
+                Moniker = GetMoniker(),
+                MaxHeight = lineHeight,
+                Margin = new Thickness(1, 0, 5, 0)
+            };
+
+            var stackPanel = new StackPanel
+            {
+                Height = lineHeight,
+                Orientation = Orientation.Horizontal
+            };
+
+            stackPanel.Children.Add(image);
+            stackPanel.Children.Add(block);
+
+            var border = new Border
+            {
+                BorderBrush = format.BackgroundBrush,
+                BorderThickness = new Thickness(1),
+                Background = Brushes.Transparent,
+                Child = stackPanel,
+                CornerRadius = new CornerRadius(2),
+                // Highlighting lines are 2px buffer. So shift us up by one from the bottom so we feel centered between them.
+                Margin = new Thickness(10, top: 0, 0, bottom: 1),
+                Padding = new Thickness(1)
+            };
+
+            border.Measure(new Size(double.PositiveInfinity, double.PositiveInfinity));
+            // Need to set these properties to avoid unnecessary reformatting because some dependancy properties
+            // affect layout.
+            // TODO: Not sure if these are needed anymore since the errors are not intratextadornment tags
```

Comment:
try removing and seeing what happens?

Reply:
I don't see any layout issues, but I was also unable to repro it when I did inline hints either.

Reply:
yea, this should be fine. They aren't as heavily impacted by typing scenarios and won't need to be measured/formatted as often since diagnostics already have some built in lag.